torchaudio/functional/functional.py

```diff
@@ -1631,3 +1637,274 @@ def rnnt_loss(
         return costs.sum()
 
     return costs
+
+
+def _compute_mat_trace(input: torch.Tensor, dim1: int = -1, dim2: int = -2) -> torch.Tensor:
+    r"""Compute the trace of a Tensor along ``dim1`` and ``dim2`` dimensions.
+
+    Args:
+        input (torch.Tensor): Tensor of dimension `(..., channel, channel)`
+        dim1 (int, optional): the first dimension of the diagonal matrix
+            (Default: -1)
+        dim2 (int, optional): the second dimension of the diagonal matrix
+            (Default: -2)
+
+    Returns:
+        torch.Tensor: trace of the input Tensor
+    """
+    assert input.ndim >= 2, "The dimension of the tensor must be at least 2."
+    assert input.shape[dim1] == input.shape[dim2], "The size of ``dim1`` and ``dim2`` must be the same."
+    input = torch.diagonal(input, 0, dim1=dim1, dim2=dim2)
+    return input.sum(dim=-1)
+
+
+def _tik_reg(mat: torch.Tensor, reg: float = 1e-7, eps: float = 1e-8) -> torch.Tensor:
+    """Perform Tikhonov regularization (only modifying real part).
+    Args:
+        mat (torch.Tensor): input matrix (..., channel, channel)
+        reg (float, optional): regularization factor (Default: 1e-8)
+        eps (float, optional): a value to avoid the correlation matrix is all-zero (Default: ``1e-8``)
+
+    Returns:
+        torch.Tensor: regularized matrix (..., channel, channel)
+    """
+    # Add eps
+    C = mat.size(-1)
+    eye = torch.eye(C, dtype=mat.dtype, device=mat.device)
+    with torch.no_grad():
+        epsilon = _compute_mat_trace(mat).real[..., None, None] * reg
+        # in case that correlation_matrix is all-zero
+        epsilon = epsilon + eps
+    mat = mat + epsilon * eye[..., :, :]
+    return mat
+
+
+def compute_power_spectral_density_matrix(
+    specgram: Tensor,
+    mask: Optional[Tensor] = None,
+    multi_mask: bool = False,
+    normalize: bool = True,
+    eps: float = 1e-10,
+) -> Tensor:
+    """Compute cross-channel power spectral density (PSD) matrix.
+
+    Args:
+        specgram (Tensor): multi-channel complex-valued STFT matrix.
+            Tensor of dimension `(..., channel, freq, time)`
+        mask (Tensor or None, optional): Time-Frequency mask for normalization.
+            Tensor of dimension `(..., freq, time)` if multi_mask is ``False`` or
+            of dimension `(..., channel, freq, time)` if multi_mask is ``True``.
+            (Default: ``None``)
+        multi_mask (bool, optional): whether to use multi-channel Time-Frequency masks. (Default: ``False``)
```

Comment:
I see. I feel like users can do ``mean`` or ``median`` reduction before calling psd method, while ``multi_mask`` can be changed to use different masks for each channel. Does that make sense to you?

Reply:
I am not sure, if providing an option to multiply different masks to different channels should be supported.
With that operation, you could break the calculation of the correlation between the channels and further the implementation is `sum (mask*obs)@(mask*obs)^H`, while you have currently `sum mask*obs@obs^H`.

Until now, I have only seen papers that do that, but not a paper that shows that it is better.
It's also not difficult to do it manually.

I only wanted to list different options, that, if such an option is provided, the name should be carefully selected to avoid confusions.